components/datetime/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+/// heavy construction of `DateTimeFormat`, and then fast formatting of `DateTime` data using the instance.
+///
+/// # Examples
+///
+/// ```
+/// use icu_locale::LanguageIdentifier;
+/// use icu_datetime::{DateTimeFormat, options::style};
+/// use icu_datetime::DummyDateTime;
+/// use icu_data_provider::InvariantDataProvider;
+///
+/// let langid: LanguageIdentifier = "en".parse()
+///     .expect("Failed to parse a language identifier.");
+///
+/// let provider = InvariantDataProvider;
+///
+/// let options = style::Bag {
+///     date: Some(style::Date::Medium),
+///     time: Some(style::Time::Short),
+///     ..Default::default()
+/// };
+/// let dtf = DateTimeFormat::try_new(langid, &provider, &options.into())
+///     .expect("Failed to create DateTimeFormat instance.");
+///
+///
+/// let date_time = DummyDateTime::new(2020, 9, 1, 12, 34, 28, 0);
+///
+/// let value = dtf.format_to_string(&date_time);
+/// ```
+///
+/// This model replicates that of `ICU` and `ECMA402` and in the future will get even more pronounce when we introduce
+/// asynchronous `DataProvider` and corresponding asynchronous constructor.
+pub struct DateTimeFormat<'d> {
+    _langid: LanguageIdentifier,
+    pattern: Pattern,
+    data: Cow<'d, structs::dates::gregory::DatesV1>,
+}
+
+impl<'d> DateTimeFormat<'d> {
+    /// `DateTimeFormat` constructor which takes a selected `LanguageIdentifier`, reference to a `DataProvider` and
+    /// a list of options and collects all data necessary to format date and time values into the given locale.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_locale::LanguageIdentifier;
+    /// use icu_datetime::{DateTimeFormat, DateTimeFormatOptions};
+    /// use icu_datetime::DummyDateTime;
+    /// use icu_data_provider::InvariantDataProvider;
+    ///
+    /// let langid: LanguageIdentifier = "en".parse()
+    ///     .expect("Failed to parse a language identifier.");
+    ///
+    /// let provider = InvariantDataProvider;
+    ///
+    /// let options = DateTimeFormatOptions::default();
+    ///
+    /// let dtf = DateTimeFormat::try_new(langid, &provider, &options);
+    ///
+    /// assert_eq!(dtf.is_ok(), true);
+    /// ```
+    pub fn try_new<D: DataProvider<'d>>(
+        langid: LanguageIdentifier,
+        data_provider: &D,
+        options: &DateTimeFormatOptions,
+    ) -> Result<Self, DateTimeFormatError> {
+        let data_key = icu_data_key!(dates: gregory@1);
+        let response = data_provider.load(&DataRequest {
+            data_key,
+            data_entry: DataEntry {
+                variant: None,
+                langid: langid.clone(),
+            },
+        })?;
+        let data: Cow<structs::dates::gregory::DatesV1> = response.take_payload()?;
+
+        let pattern = data
+            .get_pattern_for_options(options)?
+            .ok_or(error::DateTimeFormatError::MissingData)?;
```

Comment:
If I read your code correctly, the only time the pattern is `None` is when both timeStyle and dateStyle are none.  That's not a MissingData error; that's a user error.

Reply:
Thanks for spotting it! I agree. I actually even think we should silently obey - if user asks for none/none, he'll get an empty pattern.